### ISSUE TITLE
[expr.prim.lambda.closure] Conversion to function pointer doesn't account for explicit object parameter

### ISSUE THREAD

**Issue by brevzin** [cplusplus/draft#5294]
Paragraphs 8 and 9 talk about what kind of function pointer a capture-less lambda is convertible to:

> ... has a conversion function to pointer to function with C++ language [linkage](https://eel.is/c++draft/dcl.link) having the same parameter and return types as the closure type's function call operator[.](https://eel.is/c++draft/expr.prim.lambda.closure#8.sentence-1)

and

> ... The conversion function template has the same invented template parameter list, and the pointer to function has the same parameter types, as the function call operator template[.](https://eel.is/c++draft/expr.prim.lambda.closure#9.sentence-2)

Those aren't quite right in the case of an explicit object parameter. Since this should probably work:

```cpp
using P = int(*)(int);

auto f = [](this P, int i) { return i; };
auto g = [](this auto, int i) { return i; };

P pf = f;
P pg = g;
```

For the non-generic case, something like this:

```diff
- having the same parameter and return types as the closure type's function call operator
+ having the same parameter types (excluding the explicit object parameter, if any) and return type as the closure type's function call operator
```

Though wording this for the generic case is trickier since we need to skip the invented template parameter of the explicit object parameter, which might hypothetically also be used by some other parameter?

**Comment by cpplearner** [comment]
IIUC the explicit object parameter could be used in the lambda-expression. E. g.

```c++
struct C {
    C(auto);
    void f();
};

auto lambda = [](this C c) { c.f(); }(); // OK?
```

So I believe that it doesn't make sense to exclude the explicit object parameter.

**Comment by cpplearner** [comment]
IIUC the explicit object parameter could be used in the lambda-expression. E. g.

```c++
struct C {
    C(auto);
    void f();
};

auto lambda = [](this C c) { c.f(); }; // OK?
```

IIUC if the conversion function excludes the explicit object parameter, then the above `lambda` would be convertible to `void(*)()`, and the resulting function pointer would point to a function that calls `c.f()` for some nonexistent `c`. I don't think this makes sense.

**Comment by jensmaurer** [comment]
Sounds like non-editorial CWG issue material to me.

**Comment by brevzin** [comment]
> Sounds like non-editorial CWG issue material to me.

Yeah, definitely. I just wasn't sure if you preferred me emailing Core or posting it here.

**Comment by cmeerw** [comment]
But I think we still need to change

> The value returned by this conversion function is the address of a function F that, when invoked, has the same effect as invoking the closure type's function call operator on a default-constructed instance of the closure type

as it's not clear what that would mean, e.g.

```
struct C {
    C(auto) { }
};

void foo() {
    auto l = [](this C) { return 1; };
    void (*fp)(C) = l;
    fp(1); // same effect as decltype(l){}() or decltype(l){}(1) ?
}
```
(`fp` should instead just be the address of the function call operator?)

**Comment by jensmaurer** [comment]
Why is `fp` a pointer to function returning `void` when the lambda returns `int`?

**Comment by jensmaurer** [comment]
CWG2561

**Comment by frederick-vs-ja** [comment]
This should have been resolved by #7099 (4596a87af91800a8c9c3e87fdbd11df1c7874819).

**Comment by jensmaurer** [comment]
Ack, thanks!
